### ISSUE TITLE
Sorting :require dependencies alphabetically.

### ISSUE THREAD

**Issue by Kaspazza** [kkinnear/zprint#310]
Hi, is there an option to sort dependencies in a ns require?

(ns my-ns 
  (:require 
    [b.c.d.e]
    [a.c.d.e]
    [a.b.c.d])
 -->
(ns my-ns 
  (:require 
    [a.b.c.d]
    [a.c.d.e]
    [b.c.d.e])

**Comment by kkinnear** [comment]
There is not presently an option to sort dependencies in an ns require.  That said, there *is* one to sort dependencies in a leiningen `project.clj` file.  So the basic capability to sort things and thus rewrite the source file does exist.  The `ns` form and it's require are close to the format of the `project.clj` file, but not close enough that I can give you some simple option map code to do it. I could probably write and test the necessary code to sort the `:require` clauses in a `ns` macro if it would be useful to you.  

The problem is considerably more challenging than the `project.clj` file dependencies, since the `ns` macro is quite complex.  Doing all of the sorting that Stuart Sierra suggests in [How to ns](https://stuartsierra.com/2016/clojure-how-to-ns.html) would be a considerable challenge.  Just doing the `:require` might be a place to start.

This stuff is hard because rewriting code with zippers is far from intuitive and it is challenging to debug, but since I've done it once (years ago, now) for the `project.clj` file, I sure I can figure it out again if it would be of value to you.

I guess I've said the "useful to you" and "of value to you" twice because I need some encouragement from you before I jump into this.  It is an interesting problem, and I'm not against doing it, but I need a bit more than just a question before I launch into it.  

Also, if you have some real-life examples they would be useful as part of the testing if you would like me to do this.  I have all of my code, of course, but I use a fairly restrictive `ns` format. I'm always surprised at how the real world is more complex than the testing examples I use or my own code. So some real world `ns` examples from you would help me (and ensure that your code, at least, would work with whatever I create).

Oh, yes, there is one thing that can make `ns` statements better right now -- `{:style :ns-justify}`.  See [this](https://cljdoc.org/d/zprint/zprint/1.2.8/doc/zprint-reference?q=%3Ans-justify#ns-justify) which gives a before and after example of `:style :ns-justify`.  I use it in my code, I think it makes a big difference.

Thanks for asking, and I'll await your response before I start any code.

**Comment by Kaspazza** [comment]
@kkinnear Thank you very much for your answer!
I've used ns-justify and I actually spent few hours before asking just reading all the options, trying to create something with fn-map ":require" and I was basing on the require-justify, and couldn't find how to do it without modifying the code itself.

That being said, it would be really important for me and my organisation Hephaistox, to do it. As order of namespaces is not only part of it being more readable, but also a big point of an ease during PR reviews, where sometimes changes are added just because of namespace change, namespace count is also quite good predicate of how good the code in the namespace is itself (if it's not to many requires, are there some outside ones etc..). So the topic in general is very important.

I can assure you that this would be an awesome feature and we would be grateful for it. We have automated workflow for our code that checks many thing in the code itself and formatting is part of it, so it will be used in automation on many projects. 

All projects in https://github.com/hephaistox will be using it + projects that are not open to the public.

So we are fully committed to check it out and test both on spot but also in future and report if something is not right.

As a feature itself, sorting alphabetically is a great first step, but some option to say how to sort would be also awesome, as generally e.g. giving priority to namespaces from our group like "automaton" would be awesome.
Definitely we would be focused on namespace itself, so we don't really care about :as, :refer etc. in the ordering.
But it should consider the deepness of the namespace so a.b.c.d should be before a.b.c.e etc.
For requires we use all kind, as we have clojure ones, js (e.g. ["@sentry/react" :as Sentry]
   ["react-router-dom"), some java stuff. But if this would work only for clojure ones it would be great and it would be anyway most of code

**Comment by kkinnear** [comment]
Thanks, that is what I needed.  A couple of questions:

1. You said: "But it should consider the deepness of the namespace so a.b.c.d should be before a.b.c.e etc.".  I think you meant "a.b.c.d should short before a.b.c.d.e"?  Otherwise I don't have any idea what you meant.
2. I was planning on simply comparing the strings "a.b.c.d" with "a.b.c.d.e" and not breaking the namespaces apart with the periods.  That might well give you the behavior in question number 1, above. I will have to figure out how to give you all of the "automaton" namespaces first. I may be able to do something with regular expressions to show which ones to put first.  That might be a later release item, though.
3. I looked over the your code base, and found the `@sentry/react`, but not `react-router-dom` or any particularly interesting (i.e. different or unusual) Java stuff.  If you have pointers to two or three files that might demonstrate those features, it would be handy get pointers to them.   I found plenty of other things to test against.

Thanks for getting back to me.  I'll see what I can do once I'm finished fixing the current bug.

**Comment by Kaspazza** [comment]
@kkinnear like it should be:
(:require [automaton-web.configuration.protocol :as web-prot]
   [automaton-core.configuration.protocol :as core-conf-prot]
   [automaton-web.configuration.files :as web-conf-files])
->
(:require 
   [automaton-core.configuration.protocol :as core-conf-prot]
   [automaton-web.configuration.files :as web-conf-files]
   [automaton-web.configuration.protocol :as web-prot])
   
   so automaton-core is first as core is starting with 'c' so it's before 'w'
   automaton-web.configuration.files is second because files 'f' is before 'p' protocol
   Like alphabetically, but also considering sorting name after the first '.' if namespaces are from the same directories.
   
  So the point 2. is right, like simply comparing it with strings would be enough for that. And regular expression would be 100% enough for the comparison for default stuff, so like if you compare 'a' with 'b', matching with regular expression only for one of them will give them priority in sorting.
  
  3. [react-router-dom](https://github.com/hephaistox/automaton-web/blob/main/src/cljs/automaton_web/log/tracking/sentry.cljs#L5) is here. 
  
  Thank you very much, if you would need any help after or during work, let me know, I'm using zprintrc as an API to call zprint/zprint-str for automated testing over whole code, but also I have it integrated in my emacs for on-hand work. But if you want me to test on something else I can set it up oo

**Comment by Kaspazza** [comment]
Btw our config is this for now:
{:style [:community
         :how-to-ns
         :require-justify
         :sort-dependencies
         :backtranslate
         :hiccup]
 :map {:comma? false
       :force-nl? true}
 :pair {:force-nl? true}
 :pair-fn {:hang? false}
 :comment {:wrap? false}}

**Comment by Kaspazza** [comment]
@kkinnear If you need some clarification or some testing I'm here :)

**Comment by Kaspazza** [comment]
Thank you for the great work @kkinnear! 
Feature wise this is everything we need. Nice that you thought about having many regexes!
About the comments, not sorting when there are comments will be consistent with other sorting in zprintrc, as for deps.edn if you sort it with zprintrc the ones with the comment are not sorted. So in that case it's better to be consistent.
And as you, we don't leave comments in requires, because you can always use the ns docstring for that. So there is no need for that.

**Comment by arichiardi** [comment]
Oh this is an awesome feature, it will avoid my many emacs function calls to `clojure-sort-ns` :D

**Comment by Kaspazza** [comment]
@kkinnear can we test those changes on some branch? :D

**Comment by kkinnear** [comment]
Well, building the executables isn't all that straightforward, and I don't push it until I'm about to release and have done all of the testing.  I'm essentially finished with this release, and have some docs and tests to write, but then I'll release it.  It will still be a week or two at best, as I'm caught up with a bunch of other things at present.

That said, I might be able to get you an executable to try if I knew which platform you want to run it on.  Assuming that, indeed, you use one of the pre-built executables.  Do you use the Linux or macOS version?  If macOS, the Intel or Apple Silicon version?

**Comment by Kaspazza** [comment]
@kkinnear I'm okay with waiting two weeks, I don't want to put any pressure on you. It's just that you said before that you need my engagement for you to know it's worth to work on it, so I'm engaging :D 

I'm using mac m2 so apple silicon

**Comment by kkinnear** [comment]
Oh, I got the engagement!  I'm sorry I didn't make that clear.  It still wouldn't hurt for you to test it before it is released.  Let me see what I can do in the next few days.  Thanks!

**Comment by kkinnear** [comment]
I created a binary: `zprintma-1.2.9-alpha1`, and stuck it in an old alpha release place to save work: [zprint-filter-1.2.5-alpha1](https://github.com/kkinnear/zprint/releases/tag/1.2.5-alpha1).    This is an Apple Silicon binary.

I would be very interested in any testing you choose to do with this.   To get the feature that you requested, it is `{:style :sort-require}`.  It will compose well with `:style :require-justify` in case you use that too: `{:style [:sort-require :require-justify]}`.  In general, I'd put `:sort-require` first when listing a bunch of styles in a vector, though it may not make any difference.

You may also notice some slight differences in formatting in general (i.e., outside of requires) if you use reader-conditionals (i.e., `#?(:clj ... :cljs ...)`.  I fixed up the way `let` bindings are handled inside of reader-conditionals so that they still (should) sort as pairs.  It made a small difference to how two of my functions formatted in my largest file, so it isn't a major change.

The new style `:sort-require` also sorts the things in a `:refer` vector.  You can turn that off if you don't like it.  Let me know if you care and I'll tell you how to turn it off.

As you asked, you can tell zprint to sort some things in the `:require` to the top (or, as it happens, the bottom).  There are examples earlier in this issue that demonstrate how to use the `:regex-vec` to specify what comes first (or last) using a regular expression.

Have at it, and let me know how it goes, if you would.  I would love the additional testing!

Thanks for asking!

**Comment by Kaspazza** [comment]
Sorry for a late response, I had some small problems, with using that binary on mac, but it's all good now. 
(I had to do chmod +x on the file and click on it with right click -> open, before it allowed me to execute it)
So I've downloaded the binary and I've run it with command:
./zprintma '{:style [:community
         :how-to-ns
         :sort-require
         :require-justify
         :sort-dependencies
         :backtranslate
         :hiccup] :map {:comma? false
       :force-nl? true}
 :pair {:force-nl? true}
 :pair-fn {:hang? false}
 :comment {:wrap? false}}' -w **/*.clj 

It gave me a message
Failed to format file: cust_app/landing/env/development/src/clj/landing/repl.clj because java.lang.Exception:  When :list called an option-fn named 'sort-reqs' it failed because: java.lang.NullPointerException

The file content is:
```
(ns landing.repl
  "This is the entry point for starting a REPL for landing project

  Design decision:
  * All features specific to development environment should be there"
  (:require
   [automaton-core.log :as core-log]
   [automaton-core.repl :as core-repl]
   [landing.core :as landing]))

(defn- force-option?
  [args]
  (filter some? (map #(contains? #{"-f" "--force"} %) args)))

(defn -main
  "Launch repl and init landing app"
  [& args]
  (try (when-not (force-option? (first args)) (landing/-main))
       (core-repl/start-repl)
       (ns user
         (:require
          [automaton-core.dev :refer :all]))
       (catch Exception e
         ;; The print is done on purpose, as it is the message in the console if you launch the repl and it doesn't work
         (prn (str "error: " e))
         (core-log/error (ex-info "Failed to start landing, relaunch with user/"
                                  {:error e}))
         (throw e))))
```

And when I've run the same command but for *.cljc files
I've got errors in:
```
Failed to format file: automaton/automaton_core/src/cljc/automaton_core/utils/type_arg.cljc because java.lang.Exception:  When :list called an option-fn named 'sort-reqs' it failed because: clojure.lang.ArityException: Wrong number of args (0) passed to: clojure.core/max
Failed to format file: automaton/automaton_core/src/cljc/automaton_core/utils/uuid_gen.cljc because java.lang.Exception:  When :list called an option-fn named 'sort-reqs' it failed because: clojure.lang.ArityException: Wrong number of args (0) passed to: clojure.core/max
Failed to format file: automaton/automaton_core/target/la/class/automaton_core/utils/type_arg.cljc because java.lang.Exception:  When :list called an option-fn named 'sort-reqs' it failed because: clojure.lang.ArityException: Wrong number of args (0) passed to: clojure.core/max
Failed to format file: automaton/automaton_core/target/la/class/automaton_core/utils/uuid_gen.cljc because java.lang.Exception:  When :list called an option-fn named 'sort-reqs' it failed because: clojure.lang.ArityException: Wrong number of args (0) passed to: clojure.core/max
Failed to format file: automaton/automaton_core/target/production/class/automaton_core/utils/type_arg.cljc because java.lang.Exception:  When :list called an option-fn named 'sort-reqs' it failed because: clojure.lang.ArityException: Wrong number of args (0) passed to: clojure.core/max
Failed to format file: automaton/automaton_core/target/production/class/automaton_core/utils/uuid_gen.cljc because java.lang.Exception:  When :list called an option-fn named 'sort-reqs' it failed because: clojure.lang.ArityException: Wrong number of args (0) passed to: clojure.core/max
```
Which files you can see here in: https://github.com/hephaistox/automaton-core

**Comment by Kaspazza** [comment]
I'm still doing the tests, but this is error messages we've got

**Comment by kkinnear** [comment]
Wow, thank you so much doing this testing (and suggesting it).  It will be great to fix what appears to be two bugs before I release it.  I didn't expect you to find anything, so it is great to encounter these now!

**Comment by kkinnear** [comment]
Thanks again.  Some significant oversights in my testing.  `(ns stuff (:require))` failed (those were the two files you give links to).  So did `(ns stuff (:require [this :refer :all]))`.  I hadn't allowed for `:all` -- didn't actually know you could do that.

Anyway, if you want to give it another try, I put a fixed `zprintma-1.2.9-alpha2` in the same place as the `-alpha1` is. I'll understand if you don't want to bother.  But I sure hope you do, since it will make sure that this will work for you when I release it soon-ish.

Thanks again and again!

**Comment by Kaspazza** [comment]
No problem! I've tested a new binary, the errors are not appearing now and the namespaces are well sorted, both with and without the regex. It is still being tested by our team, but initially it is working well! 

I've also created a PR with update of MacOs instructions how to run binary as I've had troubles to run it in the beginning

**Comment by Kaspazza** [comment]
Hi, we've found new issue, not related to sorting, but to :require-justify

In this file the requires are not aligned with white-spaces
https://github.com/hephaistox/automaton-build/blob/main/src/clj/automaton_build/file_repo/clj_code.clj

**Comment by kkinnear** [comment]
Ah yes, you have discovered the wonders of variance based justification.   

Please bear with me as I explain this...

Some time ago, someone asked me to fix situations where there were a bunch of left-hand-side (LHS) things that were all about the same size, but one was way bigger.  Could I just justify the things that were about the same size, and let the way bigger one just be .. way bigger?  The answer was variance based justification.  Which is now the default for justification throughout zprint -- though you can turn it off, which we will get to in a bit.

The basic idea is that, when justifying a set of things, zprint will calculate the variance of the length of the LHS of the lines to justify.  If they are below the `:max-variance`, then they will all justify.  If they are above the `:max-variance`, zprint will try to leave out the longest and try again.  If the LHS (minus the longest) is below the `:max-variance`, then zprint will justify them.  If the variance is still greater than the `:max-variance`, then zprint will try one more time to leave something out.  If that works, then it will justify the LHS that were not "left out".  If that doesn't lower the variance below the `:max-variance`, then zprint will not justify the set of lines.

The point of this is that you can now leave justification "on" for lots of things in code and have it work out pretty well, in that it won't really screw up your code formatting with justifying things that really just don't work as justified.  So that variance based justification is now the default.  The default variance is typically 20, as that seems to reduce the amount of bad looking justifications to near zero.

Your two examples above happen to show both situations very cleanly.  The first, where no justification happens at all, is because the variance is over the `:max-variance` of 20 even when two things are removed.  The second justifies two of the three lines, but leaves the vary long line unjustified.

As it happens, this `-alpha2` code you are testing also includes a rework of `:require-justify` to allow more easily changing the `:max-variance`.  The documentation that you can see on how to do this is currently wrong, as I just reworked these styles to make it a lot easier.  When `:require-justify` was originally created, it was not possible to "parameterize" a style invocation.  Now, with `:style-call`, it is.  So, let's see how to do that...

You can invoke the `:require-justify` style in two ways:  `:require-justify` and `{:style-call :require-justify}`.  The first way uses a `:max-variance` of 20.  The second way, using a `:style-call` map, allows you to specify the `:max-variance` when you invoke the style.  If you want to cause `:require-justify` to justify things as much as it possibly can, you can invoke it like this: `{:style-call :require-justify :max-variance 1000}`.  This will cause it to try as hard as it can to justify things.  In some cases, where it deems the results unacceptable, it will fail to justify.  But mostly it will justify things that are currently not being justified.

Here are the `ns` macros from the front of your two example files, showing how this works.  Note that when you invoke `:require-justify` like this: `{:style-call :require-justify}` it uses the default `:max-variance` for `:require-justify`, which is 20.
```
(czprint i310r {:parse-string? true :style [:sort-require {:style-call :require-justify}]})
(ns automaton-build.file-repo.clj-code
  "Repository of files associating the name of a file to its content
  This repo deal with text files and will split the file in vector of lines"
  (:require
    [automaton-build.file-repo.raw :as build-filerepo-raw]
    [automaton-build.file-repo.raw.impl :as build-raw-impl]
    [automaton-build.file-repo.text :as build-filerepo-text]
    [automaton-build.os.files :as build-files]
    [clojure.string :as str]))

(czprint i310r {:parse-string? true :style [:sort-require {:style-call :require-justify :max-variance 1000}]})
(ns automaton-build.file-repo.clj-code
  "Repository of files associating the name of a file to its content
  This repo deal with text files and will split the file in vector of lines"
  (:require
    [automaton-build.file-repo.raw      :as build-filerepo-raw]
    [automaton-build.file-repo.raw.impl :as build-raw-impl]
    [automaton-build.file-repo.text     :as build-filerepo-text]
    [automaton-build.os.files           :as build-files]
    [clojure.string                     :as str]))

(czprint i310s {:parse-string? true :style [:sort-require {:style-call :require-justify}]})
(ns automaton-build.file-repo.text-matches
  "Find matches in a text repo, it consists of lines of analyze

  We call analysis the process and analyze the result of it"
  (:require
    [automaton-build.log :as build-log]
    [automaton-build.os.edn-utils :as build-edn-utils]
    [clojure.string      :as str]))

(czprint i310s {:parse-string? true :style [:sort-require {:style-call :require-justify :max-variance 1000}]})
(ns automaton-build.file-repo.text-matches
  "Find matches in a text repo, it consists of lines of analyze

  We call analysis the process and analyze the result of it"
  (:require
    [automaton-build.log          :as build-log]
    [automaton-build.os.edn-utils :as build-edn-utils]
    [clojure.string               :as str]))
```
Now you might be saying: "Yes, but the second examples look fine.  Why all this messing around with variance based justification?".  And I would agree that for these situations, the second example (with `:max-variance` of 1000) does look fine.  The variance based justification is largely for justifying things in code, like `let` bindings, `cond` conditions, and map keys.  It makes a big difference there.  Less so for the `ns` macro stuff, in part because the `ns` macro things all start way on the left hand side of the page.  It would probably have made sense to set the `:max-variance` for `:require-justify` to be some larger number than 20.  But I didn't, and so as to not jerk people around unnecessarily, I need to leave it at 20.  But at least it is now easy for you to change it to 1000.

Note that you cannot use `{:style-call :require-justify}` prior to `1.2.9-alpha1`.  It will not work in `1.2.8`, since I just reworked the `:require-justify` style.

I hope that explains things sufficiently.  Probably more than you wanted to know, but there it is.  

Thank you for noticing the problem and telling me about it.  It could have been a bug that I had to fix, and I would really like to get them all fixed before I release `1.2.9`.  I really appreciate your help in testing this!

----------------------------------------
Additional information (which you don't need to read, as I explained it pretty completely above):

Here is some [information](https://cljdoc.org/d/zprint/zprint/1.2.8/doc/zprint-reference?q=%3Amax-variance#a-note-on-justifying-two-up-printing) from the current documentation (which is wrong about how adjust the `:max-variance` for `:require-justify`, but correct in explaining how justification works).

Here is some [more on variance](https://cljdoc.org/d/zprint/zprint/1.2.8/doc/zprint-reference?q=%3Amax-variance#max-variance-1000).

**Comment by Kaspazza** [comment]
Thank you very much for the long explanation and solution to the justification. It makes it fully clear to me now and it was interesting to learn!

We will wait for 1.2.9 and we will than use it in the production code! 
Thank you very much for developing sorting now and all of this amazing library!

**Comment by kkinnear** [comment]
`:style :sort-require` is implemented in `1.2.9`.  Thank you for working with me to make this a reality!  I have chosen to use this style (along with `:require-justify`) for most of the files in zprint itself. I use a `:regex-vec` to put things in a general order.  I like what it does -- and wanted to use it myself so that if there were any problems I'd notice them before shipping it.

I hope it make sense.  There is pretty extensive documentation in the reference manual for `:sort-require`.

Let me know if you have any problems with this, and thanks again for pushing this idea and working with me to make it happen!

**Comment by Kaspazza** [comment]
Thank you very much @kkinnear for the update! We've applied it to our code, we will keep you updated if something pops up!
